RE: Include add-ons based on the collections endpoint/criteria in the homepage shelves API

xlisachan opened the pull request:
Fixes #15614 

This pull request updates the shelves' serializer to include add-ons to shelves with a `collections` endpoint. The update to `serializers.py` populates the add-ons for collections in the API, but I am having trouble with updating `test_serializers.py`.

I am getting the error, `AttributeError: 'WSGIRequest' object has no attribute 'query_params'`, but `query_params` were not needed when updating the serializer for collections. If `query_params` are added for collections in the test though, it leads to the following error, `django.http.response.Http404: No UserProfile matches the given query.

 I'm not sure how to resolve the error, can you please review and provide guidance when you have a moment? Thank you!

xlisachan commented:
I had forgot to include a collection in `def setUpTestData`, but I'm not sure if I added it correctly in this latest push. I am getting the error `django.http.response.Http404: No Collection matches the given query.`

I also updated `user_pk` in `shelves/serializers.py` - I had it set to `settings.TASK_USER_ID` and I believe it should be `request.user.pk` instead?

@eviljeff , can you please advise when you have a moment? Thank you!

eviljeff commented:
> I also updated `user_pk` in `shelves/serializers.py` - I had it set to `settings.TASK_USER_ID` and I believe it should be `request.user.pk` instead?

no, the task user is Mozilla, the account that also "owns" the official collections.  So the reverse is getting the collection under that account.

> I had forgot to include a collection in `def setUpTestData`, but I'm not sure if I added it correctly in this latest push. I am getting the error `django.http.response.Http404: No Collection matches the given query.`

Changing it from task user would do that - it'd be looking for that collection name under the _current_ user that called the API (which, unless you authenticated, would be anonymous), rather than the Mozilla user.

xlisachan commented:
@eviljeff, `serializers.py` has been updated to set `user_pk` back to `settings.TASK_USER_ID` per your feedback.

In addition, the following changes have been made in `test_serializers.py`:
- `def setUpTestData`: updated the collection test data by adding in `author` variable,
- `def test_url_and_addons_collections`: revised to check the url and addons in the collection shelf, 
- `def test_url_and_addons_search`: I tried to make this test a little clearer by separating the two shelves with `search` as their endpoints

`test_views.py` was also updated to include the collection test data. The number of queries changed from 4 to 26, but I am not sure if 26 should be the expected number of queries. Can you please take a look at the captured queries and let me know if it looks right- [https://pastebin.com/rnq6CzG0](url)

Please review when you have a moment. Thank you!

eviljeff commented on src/olympia/shelves/tests/test_serializers.py:
```diff
@@ -78,85 +79,69 @@ def setUp(self):
+            extra_context['action'] = 'list'
+            extra_context['kwargs'] = {
+                'user_pk': self.request.user.pk,
+                'collection_slug': instance.criteria}
```
what's this extra_context for collections used for?  I still only see `request['context']` being used in the serializer.

eviljeff replied:
argh, I wrote that the wrong way around - I should have written `context['request']`.  
My question was that you're passing `action`, and `kwargs` in as extra context (or just `context` as renamed now) but neither appear to be used in the serializer code itself, so I was wondering if that was an oversight or they were used in some hidden way (if so, please add a comment in the code explaining why they need to be passed in)

xlisachan replied:
Aah, sorry it was an oversight! I had included that when I was trying to resolve the errors when running the tests and before realizing I had not added a collection in `setUpTestData`. I didn't check if it was necessary when all the tests had finally passed.

eviljeff commented on src/olympia/shelves/serializers.py:
```diff
@@ -50,6 +51,13 @@ def get_addons(self, obj):
             request.GET = tmp
             return addons
+        elif obj.endpoint == 'collections':
+            request = self.context.get('request', None)
```
nit: you don't need to specify `None` as the default with a dict `.get` call in python - it's the default anyway.

eviljeff commented on src/olympia/shelves/tests/test_views.py:
```diff
@@ -100,7 +107,7 @@ def test_only_enabled_shelves_in_view(self):
         # don't enable shelf_c
 
-        with self.assertNumQueries(4):
+        with self.assertNumQueries(26):
```
😨 22 extra queries?  That's all from a single collection? (with one addon and a sole author)?

xlisachan replied:
I just realized the link above that outlined the captured queries did not go to [pastebin](https://pastebin.com/rnq6CzG0), sorry! I was looking through the extra queries, but I wasn't sure which were necessary and/or where it needs to be revised to correct it.

eviljeff replied:
coincidentally, this week we started looking at the amazing (in a bad way!) number of queries the collection endpoints make (#16081) so 22 queries doesn't seem so exceptional now.

xlisachan commented:
The latest push updates `extra_context` to `context` in `test/test_serializers.py` and removes the unnecessary `None` in `serializers.py`.
In addition, I corrected the link to pastebin that outlines the increase in the number of queries captured in `test/test_views.py`.
There is a conflicting file, but I'm hesitant to resolve it in case it merges this branch before you review the updates just made, as it has already been marked approved 😬🙈
Please review when you have a moment. Thanks!

xlisachan commented:
The latest push removes the unnecessary lines in `def serialize` in `test_serializers.py`. Please review when you have a moment. Thanks!

eviljeff commented on src/olympia/shelves/tests/test_serializers.py:
```diff
@@ -86,85 +85,64 @@ def setUp(self):
         )
         self.request.version = api_version
-        self.request.user = AnonymousUser()
+        self.request.user = user_factory()
```
I missed this - is it a new change?  Does AnonymousUser no longer work?  (the API should work when not logged in)

xlisachan replied:
AnonymousUser works, I'll change it back.

eviljeff reviewed (approved):
Just a question on the request user in test_serializers.py - and the branch needs a rebase like you noted.

xlisachan commented:
The latest push updates `user_factory()` back to `AnonymousUser()`. In addition, when I rebased, the lines that created the PromotedAddons in `setUpTestData` were removed, so I added them back in. Please review when you have a moment. Thank you!

eviljeff commented:
> In addition, when I rebased, the lines that created the PromotedAddons in `setUpTestData` were removed, so I added them back in. 

yeah, I did that a few weeks ago - the `recommended=True` kwarg in `addon_factory` does the same thing as manually creating the `PromotedAddon` instance in a RECOMMENDED group - in an attempt to fix intermittent test failures.  (`addon_factory` suspends the ES reindexing while it does all it's set-up so we don't end up with multiple indexing happening at once and potentially causing inconsistent data in elasticsearch).  Can you change it back.

xlisachan commented:
Thanks for the insight, I thought I hadn't rebased correctly - I have changed it back! One of the CI builds failed, but when I ran it locally it passed, can you restart it?

eviljeff reviewed (approved):
I think it's good now, r+ (!)
